Social-Music/View Controllers/ProfileViewController.m

```diff
@@ -66,19 +68,47 @@ - (void)viewDidLoad {
 - (void)viewDidAppear:(BOOL)animated {
     self.accessToken = [[SpotifyManager shared] accessToken];
     NSLog(@"access token: %@", self.accessToken);
-
-    [[TopItems shared] fetchTopData:@"artists" completion:^{
-        self.currUserArtistData = [[TopItems shared] artistData];
-        NSLog(@"%@", self.currUserArtistData);
-    }];
-    
-    [[TopItems shared] fetchTopData:@"tracks" completion:^{
-        self.currUserTrackData = [[TopItems shared] trackData];
-        NSLog(@"%@", self.currUserTrackData);
-    }];
     
-    [self.favoritesTableView reloadData];
+    PFUser *curr = PFUser.currentUser;
+
+    if (curr[@"topArtists"] == nil && curr[@"topSongs"] == nil) {
+        [[TopItems shared] fetchTopData:@"artists" completion:^{
+            self.currUserArtistData = [[TopItems shared] artistData];
+            self.currUserArtistPhotos = [[TopItems shared] artistPhotos];
+            NSLog(@"profile artist data: %@", self.currUserArtistData);
+        }];
+        
+        [[TopItems shared] fetchTopData:@"tracks" completion:^{
+            self.currUserTrackData = [[TopItems shared] trackData];
+            self.currUserTrackPhotos = [[TopItems shared] trackPhotos];
+            NSLog(@"profile track data: %@", self.currUserTrackData);
+        }];
+    } else {
+        PFQuery *trackQuery = [PFQuery queryWithClassName:@"Songs"];
+        [trackQuery whereKey:@"username" equalTo:PFUser.currentUser.username];
+        [trackQuery findObjectsInBackgroundWithBlock:^(NSArray *topSongs, NSError *error) {
+            if (topSongs != nil) {
+                self.currUserTrackData = topSongs[0][@"text"];
+                self.currUserTrackPhotos = topSongs[0][@"songImage"];
+                NSLog(@"top songs: %@", self.currUserTrackData);
+            } else {
+                NSLog(@"%@", error.localizedDescription);
+            }
+        }];
         
+        PFQuery *artistQuery = [PFQuery queryWithClassName:@"Artists"];
+        [artistQuery whereKey:@"username" equalTo:PFUser.currentUser.username];
+        [artistQuery findObjectsInBackgroundWithBlock:^(NSArray *topArtists, NSError *error) {
+            if (topArtists != nil) {
+                self.currUserArtistData = topArtists[0][@"text"];
+                self.currUserArtistPhotos = topArtists[0][@"artistImage"];
+                NSLog(@"top artists: %@", self.currUserArtistData);
+            } else {
+                NSLog(@"%@", error.localizedDescription);
+            }
+        }];
+    }
+    [self.favoritesTableView reloadData];
```

Comment:
Does this work? Seems like the reloadData is happening at the same time as the call. You are setting the data correctly above though.

Reply:
It seems to work but that may be because the call is instantaneous - I can put it in a completion block if that's better.